### PR TITLE
flow: fs node transform into persistence node - with efivars - v5

### PR DESCRIPTION
**Pull request by edersondisouza**
v5:
Instead of error packets on no previous stored value, now we use a 'default_value';
RESET port to reset node to its default_value;
As the code was reorganised, the commit about 'alloca' as droped as its content merged into previous commit.

v4:
No changes. Just to bring semaphore back.

v3:
Return error if packet requested size differs from persisted size;
Fixed make allyesconfig failure;
Added option 'error_on_no_value' - when true (default false) it'll send an error packet informing that no previous value was found upon node opening.

v2:
No more 'fs_dir_path'. If storage is 'fs', 'name' can contain a fully fledged path to a file;
Using more sol_buffer;
Using additional read and write to deal with efivars 'attributes' - avoid allocations and simplifying code;
Depending on Linux for now, thus avoiding some ifdefs;

Adding 'efivars' support to Soletta. I decided to reuse 'fs' node type, but renamed it to 'persistence' - however name suggestions are welcome.
From user's perspective, it's like a dictionary: give a name to a packet and save it. User can choose storage type, currently 'efivars' and 'fs'. In case it chooses 'fs', [s]he must choose a 'fs_dir_path' - dir where persisted files will reside - but I'm open to suggestion on how to improve this: I didn't like having to make user define it for every persistence node, but maybe is more a question of config file.

### CODE REVIEW

File: src/modules/flow/persistence/fs-storage.c [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+{
+    FILE *file;
+
+    file = fopen(name, "w+e");
+    if (!file) {
+        SOL_WRN("Could not open persistence file [%s]", name);
+        return -errno;
+    }
+
+    fwrite(buffer->data, buffer->used, 1, file);
```
Comment (lpereira) [inline src/modules/flow/persistence/fs-storage.c]: Not checking the return valud of `fwrite()`: is this intentional?

File: src/modules/flow/persistence/persistence.c [hunk 2]
```diff
@@ -0,0 +1,557 @@
...
+    char *storage;
+
+    struct sol_flow_packet *(*packet_new_fn)(const struct persist_data *data);
+    int (*packet_data_get_fn)(const struct sol_flow_packet *packet, void *value_ptr);
+    int (*packet_send_fn)(struct sol_flow_node *node);
+    void *(*node_get_default_fn)(struct sol_flow_node *node);
+
+    int (*storage_write)(const char *name, struct sol_buffer *buffer);
+    int (*storage_read)(const char *name, struct sol_buffer *buffer);
+    int (*storage_get_size)(const char *name, size_t *size);
```
Comment (lpereira) [inline src/modules/flow/persistence/persistence.c]: I'm not sure if I like this interface: it's prone to TOCTOU. Would it be possible to have only primitives for reading and writing, with the read primitive returning the size or an error?

File: src/modules/flow/persistence/persistence.c [hunk 3]
```diff
@@ -0,0 +1,557 @@
...
+    struct persist_data *mdata = data;
+    size_t size;
+    int r;
+
+    if (!storage || *storage == '\0') {
+        SOL_WRN("Must define a storage type");
+        return -EINVAL;
+    }
+
+    if (streq(storage, "fs")) {
```
Comment (lpereira) [inline src/modules/flow/persistence/persistence.c]: sol_str_table?

File: src/modules/flow/persistence/fs-storage.c (hunk 1 above)
Comment (edersondisouza) [inline src/modules/flow/persistence/fs-storage.c]: Maybe I'm kind of `fwrite()` won't fail, if something goes wrong it'll arise on `fclose()`, but I can check return here too - maybe string values can be big enough to fill file buffer.

File: src/modules/flow/persistence/persistence.c (hunk 2 above)
Comment (edersondisouza) [inline src/modules/flow/persistence/persistence.c]: Hmmm... Fair. However, in this case reading will need to take care of allocating memory - but I think that `sol_buffer` can be of help.

File: src/modules/flow/persistence/persistence.c (hunk 3 above)
Comment (edersondisouza) [inline src/modules/flow/persistence/persistence.c]: Didn't know about it, will check.

**Comment by edersondisouza** [comment]
v6 on #666

**Pull request closed without merging**